### ISSUE TITLE
Detected concurrency problem with Query Gating

### ISSUE THREAD

**Issue by Jag84** [sebastienros/yessql#444]
There's a problem while trying to save two changes on the same Document concurrently.

I've simplify the issue into two tests as you can see in this PR at my YesSQL fork. The fork is  created now for showing you the problem:

https://github.com/Jag84/yessql/pull/1

If you have Query Gating enabled and you try to update a document concurrently, things couldn't work as it is expected. It seems sometimes document stored contains changes from another not commited changes. It causes that a retry operation attemps to remove an element which failed due to concurrent exception and it fails cause the element doesn't exist (despite its original delete operation failed due to a concurrent exception).

If you disable Query Gating, everything works fine.

If you want to check it by yourself, you only need to clone and run both `ShouldUpdateConcurrentDocumentsWithQueryGating` and `ShouldUpdateConcurrentDocumentsWithoutQueryGating` tests.

**Comment by Jag84** [comment]
I can solve the failure removing the element that doesn't exist returning a copy of the Document on `ProduceAsync`. The changed method would be something like this: 

```
  internal async Task<T> ProduceAsync<T, TState>(WorkerQueryKey key, Func<TState, Task<T>> work, TState state)
  {
      if (!Configuration.QueryGatingEnabled)
      {
          return await work(state);
      }

      object content = null;

      while (content == null)
      {
          // Is there any query already processing the ?
          if (!Workers.TryGetValue(key, out var result))
          {
              // Multiple threads can potentially reach this point which is fine
#if !NET451
              // c.f. https://blogs.msdn.microsoft.com/seteplia/2018/10/01/the-danger-of-taskcompletionsourcet-class/
              var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
#else
              var tcs = new TaskCompletionSource<object>();
#endif

              Workers.TryAdd(key, tcs.Task);

              try
              {
                  // The current worker is processed
                  content = await work(state);
              }
              catch
              {
                  // An exception occurred in the main worker, we broadcast the null value
                  content = null;
                  throw;
              }
              finally
              {
                  // Remove the worker task before setting the result.
                  // If the result is null, other threads would potentially
                  // acquire it otherwise.
                  Workers.TryRemove(key, out result);

                  // Notify all other awaiters to return the result
                  tcs.TrySetResult(content);
              }
          }
          else
          {
              // Another worker is already running, wait for it to finish and reuse the results.
              // This value can be null if the worker failed, in this case the loop will run again.
              content = await result;

              // CREATED CONTENT COPY HERE
              if (!(content is int))
              {
                  content = Configuration.ContentSerializer.Deserialize(Configuration.ContentSerializer.Serialize(content), content.GetType());
              }
          }
      }

      return (T)content;
  }
}
```

Even so, the concurrency problem keeps happening.

**Comment by jersiovic** [comment]
It has sense that query gating can produce unexpected problems with two threads reading same document from the db if both want to modify it, reason is they will share same object in memory and both can apply changes to the same object. So, no matter if one thread receives a concurrency exception storing object in db because the other will succeed storing it in db and that one will store the shared object that will contain both modifications.

So, it seems sensible what you did on your second comment to create a clone of the document to return for those workers waiting a result. And for sure this has fixed some problems. What I don't get is what else could be causing that after that at the end of the test not all de nationalities in the document has been removed

**Comment by jersiovic** [comment]
I've found time to come back to this and I've found what is causing the bug. 
Problem is the first thread adding an object in Workers for a key is sharing with others a reference to result that it is returned by the method. So, this thread after returning the object changes the object that is the same other threads are reciving. So, no matter if the serialize and deserialize it to work with a new copy of the object. Problem is the original object they receive has been changed and it is not the same in database.

If you want to provide me write access to your repo I can make there changes to your pr to make it work
